x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/rules/action/PutQueryRulesetAction.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.rules.action;
+
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.StatusToXContentObject;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.application.rules.QueryRuleset;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public class PutQueryRulesetAction extends ActionType<PutQueryRulesetAction.Response> {
+
+    public static final PutQueryRulesetAction INSTANCE = new PutQueryRulesetAction();
+    public static final String NAME = "cluster:admin/xpack/query_rules/put";
+
+    public PutQueryRulesetAction() {
+        super(NAME, PutQueryRulesetAction.Response::new);
+    }
+
+    public static class Request extends ActionRequest {
+
+        private final QueryRuleset queryRuleset;
+        private final boolean create;
+
+        public Request(StreamInput in) throws IOException {
+            super(in);
+            this.queryRuleset = new QueryRuleset(in);
+            this.create = in.readBoolean();
+        }
+
+        public Request(QueryRuleset queryRuleset, boolean create) {
+            this.queryRuleset = queryRuleset;
+            this.create = create;
+        }
+
+        public Request(String rulesetId, boolean create, BytesReference content, XContentType contentType) {
+            this.queryRuleset = QueryRuleset.fromXContentBytes(rulesetId, content, contentType);
+            this.create = create;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+
+            // TODO: validate query ruleset
```

Comment:
I'm not quite sure what validation we'd want here that we don't already do in constructors. Ideas welcome?

Reply:
I can't think of any! Maybe we can change the TODO to a comment indicating that they are already validated in constructors

Reply:
So, I would actually try to move all the validations here if possible.
Because right now, what happens is that we validate the request payload before we validate the license.
We want the license check to be the first one that happens, not the validation of the payload.
It's a problem we have with the rest of the APIs in this module, that we did not get to fix (we have an issue for it).
But since this is a new API, maybe we can make this right from the beginning?

Reply:
Updated with validations here. Is that enough, @ioanatia? I don't like the idea of removing constructor validation too, but we can discuss for a followup PR